Docs - testing of integration guides (#114)
kgateway-dev/kgateway.dev: content/docs/integrations/external-dns-cert-manager.md

---
title: ExternalDNS and cert-manager
weight: 520
---

[ExternalDNS](https://github.com/kubernetes-sigs/external-dns) and [cert-manager](https://github.com/cert-manager/cert-manager) are two well-known integrations within the Kubernetes ecosystem that can be used in conjunction to automate the creation of TLS certificates. 

> * **ExternalDNS**: Instead of manually editing your domain in your DNS provider to add load balancer IP addresses, you can use ExternalDNS to dynamically set up and control DNS records for discovered gateway and HTTP resources. When you define a hostname in an HTTPRoute resource, ExternalDNS uses the external address that is assigned to the gateway's load balancer service that serves this hostname, and uses this information to create a DNS record in the DNS provider that you configured.

Reviewer: ```suggestion
* **ExternalDNS**: Instead of manually editing your domain in your DNS provider to add load balancer IP addresses, you can use ExternalDNS to dynamically set up and control DNS records for discovered Gateway and HTTPRoute resources. When you define a hostname in an HTTPRoute resource, ExternalDNS uses the external address that is assigned to the Gateway's load balancer service that serves this hostname, and uses this information to create a DNS record in the DNS provider that you configured.
```

> * **cert-manager**: You can then use cert-manager to quickly and programmatically generate certificates for your domain, and store them in a Kubernetes secret. By adding this certificates secret to the HTTPRoute on the gateway for your domain, you can secure it for HTTPS traffic.

Reviewer: ```suggestion
* **cert-manager**: You can then use cert-manager to quickly and programmatically generate certificates for your domain, and store them in a Kubernetes secret. By adding this certificates secret to the HTTPRoute on the Gateway for your domain, you can secure it for HTTPS traffic.
```

## Before you begin 

{{< reuse "docs/snippets/prereq.md" >}}

## Set up ExternalDNS

> Use ExternalDNS to dynamically set up and control DNS records for discovered gateway and HTTP resources. When you define a hostname in an HTTPRoute resource, ExternalDNS uses the external address that is assigned to the gateway's load balancer service that serves this hostname, and uses this information to create a DNS record in the DNS provider that you configured.

Reviewer: ```suggestion
Use ExternalDNS to dynamically set up and control DNS records for discovered Gateway and HTTPRoute resources. When you define a hostname in an HTTPRoute resource, ExternalDNS uses the external address that is assigned to the gateway's load balancer service that serves this hostname, and uses this information to create a DNS record in the DNS provider that you configured.
```

Reviewer: do we need this paragraph here?

> In the next section, you use cert-manager to create TLS certificates for this hostname so that you can serve HTTPS traffic on your gateway. 

Reviewer: ```suggestion
In the next section, you use cert-manager to create TLS certificates for this hostname so that you can serve HTTPS traffic on your Gateway. 
```

   - kind: ServiceAccount
     name: external-dns
     namespace: default
   EOF
   ```
   

> 4. Create the deployment for ExternalDNS, which configures ExternalDNS to monitor gateway and HTTP route resources to determine the list of DNS records that must be created or changed. Note that in the following example, DNS records are set up in DigitalOcean, in which you provide your token. If you use a different DNS provider, find the required ExternalDNS configuration settings in the [Kubernetes documentation](https://kubernetes-sigs.github.io/external-dns/v0.14.0/#deploying-to-a-cluster).

Reviewer: ```suggestion
4. Create the deployment for ExternalDNS, which configures ExternalDNS to monitor Gateway and HTTPRoute resources to determine the list of DNS records that must be created or changed. Note that in the following example, DNS records are set up in DigitalOcean, in which you provide your token. If you use a different DNS provider, find the required ExternalDNS configuration settings in the [Kubernetes documentation](https://kubernetes-sigs.github.io/external-dns/v0.14.0/#deploying-to-a-cluster).
```

```

## Cleanup

{{< reuse "docs/snippets/cleanup.md" >}}

1. Remove the cert-manager annotation and HTTPS listener from the `http` Gateway.
   ```yaml
   kubectl apply -f- <<EOF
   kind: Gateway
   apiVersion: gateway.networking.k8s.io/v1
   metadata:
     name: http
     namespace: {{% reuse "docs/snippets/ns-system.md" %}}
   spec:
     gatewayClassName: kgateway
     listeners:
     - protocol: HTTP
       port: 8080
       name: http
       allowedRoutes:
         namespaces:
           from: All
   EOF
   ```

2. Delete the Issuer resource.
   ```sh
   kubectl delete Issuer letsencrypt-http -n {{< reuse "docs/snippets/ns-system.md" >}}
   ```

3. Delete the `letsencrypt-http-issuer-account-key` secret.
   ```sh
   kubectl delete secret letsencrypt-http-issuer-account-key -n {{< reuse "docs/snippets/ns-system.md" >}}
   ```

4. Remove cert-manager.
   ```sh
   kubectl uninstall cert-manager -n cert-manager
   kubectl delete ns cert-manager
   ```

5. Delete the ExternalDNS deployment and related resources.
   ```sh
   kubectl delete Deployment external-dns
   kubectl delete ClusterRoleBinding external-dns
   kubectl delete ClusterRole external-dns

>    kubectl delete ServiceAccount external-dns

Reviewer: can add deleting the namespace

Author: It installs into default
